Dropdown select limit (view limit) option for Paginator

eather009 opened the pull request:
Dropdown array for Select limit

@param array $limits This is array of options.
@param int|null $default Default limit for option selecting. Default value is $this->param('perPage').
@param array $options Options for Select tag attributes like class, id or event
@return string html output.

This will help to show by data limit for per page without writing extra code, just add following code will give you below image output.

`$this->Paginator->limitControl()`

<img width="109" alt="limit_control" src="https://cloud.githubusercontent.com/assets/4234756/23317055/1d6b9200-faf7-11e6-9b78-78c90c8f8971.png">

Also override select tag attribute can be possible by passing $options.

Thank you

eather009 committed:
Dropdown select limit (view limit) option for Paginator

eather009 committed:
select all option added

eather009 committed:
change empty quotation to 0

eather009 committed:
formation update

eather009 committed:
formation update

eather009 committed:
Merge branch '3.next' of https://github.com/eather009/cakephp into 3.next

# Conflicts:
#	src/View/Helper/PaginatorHelper.php

eather009 committed:
short array defined

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,37 @@ public function implementedEvents()
+        }
+
+        if (empty($limits)) {
+            $limits += ['0' => __('All'), '20' => '20', '50' => '50', '100' => '100'];
```
0 limit makes no sense.

eather009 replied:
Im getting error 
codecov/project — 94.9% (-0.14%) compared to 4180750 and 
codecov/patch — 91.66% of diff hit (target 95.03%) but I don't know why. Thus set 0 to check whether it has any dependency or not.

ADmad replied:
That's probably because you haven't added any tests case calling the method without any params.

eather009 replied:
Thank you. Ya, that was my missing.

eather009 committed:
remove 0 limit

eather009 committed:
test case handle with default limit params

ravage84 commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,41 @@ public function implementedEvents()
+     * @param array $limits This is array of options.
+     * @param int|null $default Default limit for option selecting. Default value is $this->param('perPage').
+     * @param array $options Options for Select tag attributes like class, id or event
+     *
```
No white line.

eather009 replied:
Thank you. Given.

ravage84 commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,41 @@ public function implementedEvents()
     }
+
+    /**
+     * dropdown array for Select limit
```
Dropdown select for pagination limit

eather009 replied:
Thank you. Updated.

ravage84 commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,41 @@ public function implementedEvents()
+    /**
+     * dropdown array for Select limit
+     *
+     * @param array $limits This is array of options.
```
The options array.

eather009 replied:
Thank you. Given.

ravage84 commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,41 @@ public function implementedEvents()
+     * dropdown array for Select limit
+     *
+     * @param array $limits This is array of options.
+     * @param int|null $default Default limit for option selecting. Default value is $this->param('perPage').
```
`````
Default option for pagination limit. Defaults to ``$this->param('perPage')``.

ravage84 commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,41 @@ public function implementedEvents()
+        }
+
+        if (empty($limits)) {
+            $limits = [
```
Since these values can vary greatly between applications,, I'm not sure if it makes much sense to set such fallbacks.

dereuromark replied:
Some defaults are OK IMO.

eather009 replied:
Users can set any options, even they may want keep it default, which will help user to get default options. So, this condition is given.

ravage84 commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,41 @@ public function implementedEvents()
+            ];
+        }
+
+        if (!in_array($default, $limits)) {
```
Does it make sense to be able to set the default to a value that is not a valid option?

eather009 replied:
Actually, this have another usage. I am using cache for pagination limit which will stored each page pagination limit value. Now, if anyone wants his own value which is not included in the Limit options, that will be added to the limits. 

Yes, in one sense, this really doesn't make sense here unless anyone using caching or other technique on pagination

eather009 replied:
Thank you. Updated.

ravage84 commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,41 @@ public function implementedEvents()
+            $limits += [$default => $default];
+        }
+
+        ksort($limits);
```
Not sure everybody wants to have it sorted this way. I could imagine some want to have a poplular value on top of the list.

eather009 replied:
The reason is same as before I mentioned in previous comment. Actually, this was used for my own purpose. I can remove this. Thank you

eather009 replied:
Thank you. Updated.

ravage84 commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,41 @@ public function implementedEvents()
+
+        ksort($limits);
+
+        $out .= $this->Form->control('limit', ($options + ['type' => 'select', 'label' => __('View'), 'value' => $default, 'options' => $limits, 'onChange' => 'this.form.submit()']));
```
Line too long.
What's with that bracket before $options?
May be put that into a separate assignment before assembling the output.

eather009 replied:
Thank you. Updated.

ravage84 commented on tests/TestCase/View/Helper/PaginatorHelperTest.php:
```diff
@@ -2546,4 +2546,78 @@ public function testMeta($page, $prevPage, $nextPage, $pageCount, $options, $exp
+        ];
+        $this->assertHtml($expected, $out);
+
+        $out = $this->Paginator->limitControl([], null, ['class' => 'form-control']);
```
Setting the default option is not tested.

eather009 replied:
Thank you. Updated.

eather009 committed:
Comments updated
Removed sort and set defaults to limit if not in array checking
Break $out options

markstory commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,40 @@ public function implementedEvents()
     }
+
+    /**
+     * Dropdown select for pagination limit
```
Might be good to mention that this will also generate the wrapping form as I am sure someone will put this inside an already open search form.

eather009 replied:
Ya. Sure. Giving update

eather009 replied:
Thank you. Its given

eather009 committed:
Notify that limitControl will generate a wrapping Form

dereuromark commented:
Should we accept and merge this? Is everything in order now?
